torch/csrc/jit/codegen/cuda/runtime/block_welford_outer.cu

```diff
@@ -0,0 +1,265 @@
+// blockDim.x as the reduction is done along threadIdx.y. So to be
+// effective, blockDim.x needs to be smaller than WARP_SIZE, and in the
+// case of grouped grid welford, it should be typically 8 or 16.
+//
+// The algorithm is an adaptation of scattered butterfly reduction,
+// aka recursive halving, commonly used for implementing
+// MPI_Reduce_scatter. For a visual illustration of the data
+// organization, see, for example, page 22 of Solomonik,
+// Design of Parallel and High-Performance Computing:
+// Distributed-Memory Models and Algorithms, 2015
+// (https://solomonik.cs.illinois.edu/talks/dphpc-dec-2015.pdf)
+//
+// Assumptions:
+// - blockDim.x and blockDim.y are statically known values so that all
+// loops can be completely unrolled
+// - blockDim.x is smaller than WARP_SIZE
+// - blockDim.x evenly divides WARP_SIZE
+// - There are multiple warps per block
+// - The gouping factor, NumVals, is at least as large as the warp
+// dimY and is divisible by the warp dimY.
+//
+// This is meant to be used as part of the grouped grid welford
+// reduction but should be usable as a standalone block welford routine as
+// long as the above assumptions hold.
+//
+// Note: Having an output reference parameter resulted in using more
+// registers than just returing the output. Results would vary
+// depending on compiler versions, but it seems safer to return outputs
+// as a new value.
+template <int NumVals, typename DataType, int BDIMX, int BDIMY>
+__inline__ __device__ WelfordTriplet<DataType> blockWelfordOuter(
+    DataType* inp_avg,
+    DataType* inp_var,
+    nvfuser_index_t inp_N,
+    DataType* smem) {
+  constexpr int num_warps = BDIMX * BDIMY / 32;
+  static_assert(num_warps >= 1, "There must be at least a single warp");
+  static_assert(32 % BDIMX == 0, "blockDimx.x must be able to divide 32");
+
+  const int tid = threadIdx.x + threadIdx.y * BDIMX;
+  const int wid = tid / 32;
+
+  // Dimension of the Y axis within each warp
+  constexpr int wdimy = 32 / BDIMX;
+  static_assert(NumVals >= wdimy, "NumVals must be >= 32 / blockDim.x");
+  static_assert(
+      NumVals % wdimy == 0, "NumVals must be divisible by 32 / blockDim.x");
+  // There must be at least a single warp
+
+  // Y index within each warp
+  const int warp_tidy = threadIdx.y % wdimy;
+
+  // Thread index in each warp
+  const int lane_id = threadIdx.x + warp_tidy * BDIMX;
+
+  constexpr int smem_var_offset = num_warps * BDIMX * NumVals;
+  constexpr int smem_N_offset = num_warps * BDIMX * NumVals * 2;
+
+  // We define a chunk as a value in a group and a chunk size as the
+  // number of group values per thread. Initiall, the chunk size is
+  // NumVals. After the initial warp reduction, the chunk size is
+  // reduced to NumVals/wdimy. For example, support NumVals=8,
+  // blockDim.x=8, blockDim.y=32, then wdimy=4, so afer the initial
+  // warp reduction, the chunk size is 2. This is the number of
+  // elements each thread stores to shared memory.
```

Comment:
Are you moving threads in the y dimension to help in the x dimension as they get reduced out?

Reply:
Not sure which dimension is the x dimension, but threads in the y dimension divide a group of `NumVals`.